kratos/python_scripts/analysis_stage.py

```diff
@@ -127,12 +168,96 @@ def ModifyInitialGeometry(self):
     def ApplyBoundaryConditions(self):
         """here the boundary conditions is applied, by calling the InitializeSolutionStep function of the processes"""
 
-        for process in self.list_of_processes:
+        for process in self._GetListOfProcesses():
             process.ExecuteInitializeSolutionStep()
 
         #other operations as needed
 
-
     def ChangeMaterialProperties(self):
         """this function is where the user could change material parameters as a part of the solution step """
         pass
+
+    def _GetSolver(self):
+        if not hasattr(self, '_solver'):
+            self._solver = self._CreateSolver()
+        return self._solver
+
+    def _CreateSolver(self):
+        """Create the solver
+        """
+        raise Exception("Creation of the solver must be implemented in the derived class.")
+
+    def _GetListOfProcesses(self):
+        """This function returns the list of processes involved in this Analysis
+        The list of processes is constructed in case it is not existing yet
+        """
+        if not hasattr(self, '_list_of_processes'):
+            order_processes_initialization = self._GetOrderOfProcessesInitialization()
+            self._list_of_processes = self._CreateProcesses("processes", order_processes_initialization)
+            list_output_processes = self._GetListOfOutputProcesses()
+            self._list_of_processes.extend(list_output_processes) # Adding the output processes to the regular processes
+        return self._list_of_processes
+
+    def _GetListOfOutputProcesses(self):
+        """This function returns the list of output processes involved in this Analysis
+        The list of output processes is constructed in case it is not existing yet
+        """
+        if not hasattr(self, '_list_of_output_processes'):
+            order_processes_initialization = self._GetOrderOfOutputProcessesInitialization()
+            self._list_of_output_processes = self._CreateProcesses("output_processes", order_processes_initialization)
+        return self._list_of_output_processes
+
+    def _CreateProcesses(self, parameter_name, initialization_order):
+        """Create a list of processes
+        Format:
```

Comment:
This seems too much complex to me. Is it required by any technical reason?

Reply:
The idea behind grouping the processes is that one can loop them. 
this way we can solve #869 and also automatize the computation of the `ComputingModelPart` (which is quite error-prone)

Without the extra level it would not be possible to specify the order of execution (like it is required e.g. in Fluids)

Users can still add their custom processes, but it should be done under `processes` or `output_processes` respectively

Reply:
Go ahead then!